### PR TITLE
BFD-1641: RFC for Centralized Config Management

### PR DESCRIPTION
**Pull request by mjburling**
~**This cannot be merged without the approval of @StewGoin**~ Approved! Thanks again, @StewGoin!

**JIRA Ticket:**
[BFD-1641](https://jira.cms.gov/browse/BFD-1641)

**User Story or Bug Summary:**

RFC for centralized configuration management store.

---

### What Does This PR Do?

RFC proposes adoption of AWS SSM Parameter Store and conventions for managing configuration data therein.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [x] Yes
      * [ ] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: rfcs/0000-centralized-configuration-management.md [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+
+We take advantage of Parameter Store's hierarchical nature by formatting the hierarchies into 4-tuple prefixes, culminating in a named _leaf_ node. Together, the 4-tuple prefix and leaf make up a single parameter: `/${project}/${env}/${group}/${sensitivity}/${leaf}`.
+Each member of the tuple represents a scoped level or tier providing increasingly specific context for BFD services and supporting systems.
+The scopes lend themselves well to crafting simple queries and tailored access policies that target the necessary hierarchies of configuration data only.
+The primary use cases for these parameters include infrastructure-as-code, automated-configuration-management, and release-management strategies in `terraform`, `ansible`, and `jenkins` tooling, respectively.
+This proposed format **does not** apply to other potential Parameter Store use-cases.
+
+### Project Scope
+
+Project scope (`${project}`) is the _parameter name root_. In these hierarchies, this is *practically* an invariant that resolves to our project: `bfd`.
```
Comment (mackec1) [inline rfcs/0000-centralized-configuration-management.md:81]: Question: does paca live within bfd or it's own root?

File: rfcs/0000-centralized-configuration-management.md [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+#### Examples
+
+- `/bfd/mgmt/jenkins/`
+- `/bfd/prod-sbx/migrator/`
+- `/bfd/prod/server/`
+- `/bfd/test/common/`
+
+### Sensitivity Scope and the Leaf Node
+
+Sensitivity scope (`${sensitivity}`) defines whether a leaf node (`${leaf}`) in the hierarchy is stored as an ordinary `String` or encrypted with AWS KMS as a `SecureString`.
```
Comment (mackec1) [inline rfcs/0000-centralized-configuration-management.md:124]: As far as KMS, will there be a single key or multiple keys predicated on the env scope?

**Review by mackec1 — APPROVED** [review]
REALLY GOOD explanation of benefits, constraints, etc. Best RFC I've read in a hwile!

File: rfcs/0000-centralized-configuration-management.md [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+Instead, the sprawling configuration has historically discouraged us from adopting worthy changes, contributed to operators' inadvertent addition of errors, and represents an unfriendly introduction to new contributors who wish to make what would otherwise be simple updates to BFD configuration.
+
+## Proposed Solution
+[Proposed Solution]: #proposed-solution
+
+AWS Systems Manager (SSM) Parameter Store is a hierarchical, key-value data store typically used for centrally managed sensitive and nonsensitive configuration data.
+In addition to the AWS Console, the Parameter Store is accessed through the Parameter Store API, and can be even more accessible through various tools, including:
+- the AWS CLI via `aws ssm` sub-command
+- terraform resources and data sources with the aws provider
+- ansible workflows through the `community.aws.aws_ssm_parameter_store` module
```
Comment (karlmdavis) [inline rfcs/0000-centralized-configuration-management.md:69]: Wouldn't it be more `https://docs.ansible.com/ansible/2.5/plugins/lookup/aws_ssm.html`?

File: rfcs/0000-centralized-configuration-management.md (hunk 1 above)
Comment (mjburling) [reply rfcs/0000-centralized-configuration-management.md:81]: > Question: does paca live within bfd or it's own root?

We could adopt an alternative root for RDA if it makes sense. At this time, the conventional `bfd` root will do.

**Review by christrotteradhoc — DISMISSED** [review]
Great work!

**Review by christrotteradhoc — APPROVED** [review]

**Pull request merged**